cranelift-codegen/src/peephole.rs

```diff
@@ -0,0 +1,469 @@
+            domtree,
+        };
+        ctx.run()
+    }
+}
+
+struct Context<'a> {
+    // Current instruction as well as reference to function and ISA.
+    cur: EncCursor<'a>,
+
+    // Cached ISA information.
+    // We save it here to avoid frequent virtual function calls on the
+    // `TargetIsa` trait object.
+    reginfo: RegInfo,
+
+    // References to contextual data structures we need.
+    domtree: &'a DominatorTree,
+}
+
+impl<'a> Context<'a> {
+    fn run(&mut self) {
+        self.cleanup();
+    }
+}
+
+// ======== A description of the redundant-fill-removal algorithm ========
+//
+// The simplest incarnation simply detects fills from the same stack slot into
+// the same register, and removes them:
+//
+//  v1 = fill v0
+//  ...
+//  v2 = fill v0
+//
+// The second fill is redundant provided that (1) v1 and v2 have the same
+// register, (2) that "..." overwrites neither that register nor v0's stack
+// slot, and (3) that the second fill is dominated by the first.  (3) is
+// implied since we're doing this only inside EBBs, but we have to check (1)
+// and (2).
+//
+// In this example, the instruction processing loop below creates an
+// environment entry on encountering "v1 = fill v0", and uses that entry to
+// determine that "v2 = fill v0" is redundant.
+//
+// The loop must also, even in the simplest version, check all instructions
+// for writes to registers and stack slots, and invalidate environment entries
+// accordingly.
+//
+// === First refinement -- implemented
+//
+// This is to observe that spills also create equivalences between registers
+// and stack slot values, which in turn increases our changes of proving later
+// fills redundant.
+//
+// === Second refinement -- implemented
+//
+// For a fill, even if we can't find an exact match in the environment,
+// finding a load of the same spill slot into a different register is still a
+// win, since we can convert the fill into a reg-reg copy.
+//
+// === Other tricks that don't seem to help
+//
+// I tried tracking the effects of InstructionData::Copy in the environment,
+// for the case when both source and destination are registers.  But that
+// didn't cause any more fills to be removed or converted to copies.  I'm not
+// sure why.
+//
+// It's also possible for a spill to be detected as redundant, if such a
+// binding already appears in the environment.  I implemented this, but it
+// seems almost entirely ineffective, so I removed it again.
+
+// This describes a spill slot in the obvious way.  It also tracks the SSA
+// value name associated with the slot.
+#[derive(Clone, Copy)]
+struct SlotInfo {
+    value: Value,
+    offs: i32,
+    size: u32,
+}
+
+// Find out if the range `[offs, +size)` overlaps the with range in `si`.
+fn any_overlap(si: &SlotInfo, offs: i32, size: u32) -> bool {
+    let a_offs = si.offs as i64;
+    let a_size = si.size as i64;
+    let b_offs = offs as i64;
+    let b_size = size as i64;
```

Comment:
These could all be `u64` instead, right? (Or are offsets negative?)

Reply:
These can't be `u64`, since offsets can be negative; they are `i32`s.  I cast them to `i64` to do the checks in order to avoid potentially complex range-end / wraparound logic.